### DISCUSSION TITLE
How to bootstrap Stylet witout Application xaml

### DISCUSSION THREAD

**Discussion by mwdpb** [canton7/Stylet#206] [Q&A] [answered]
The code snippet in the documentation is for a WPF project which has Application.xaml.  However, I'd like to use Stylet in an Office VSTO add-in project.  How and where should I bootstrap Stylet?  Please advice, thank you!

**Comment by canton7** [comment]
I'm afraid I don't have any experience with Office VSTO add-ins. You don't *need* to use the ApplicationLoader, but if you don't you'll need to construct the Bootstrapper yourself (and call its `Setup` method, passing in your Application), and you won't get any of Stylet's resources unless you merge in Stylet's ResourceDictionary yourself.

**Reply by mwdpb** [reply]
What should we do in a class library project then?  The main program cannot be modified and it'll call a plug-in which is a class library.  How should you bootstrap in a class library?  It doesn't have Application xaml either.

I felt they are similar and I'm actually facing both situations, :)

**Reply by mwdpb** [reply]
> I'm afraid I don't have any experience with Office VSTO add-ins. You don't _need_ to use the ApplicationLoader, but if you don't you'll need to construct the Bootstrapper yourself (and call its `Setup` method, passing in your Application), and you won't get any of Stylet's resources unless you merge in Stylet's ResourceDictionary yourself.

For the resource part, I think we can manually add that in each xaml file, a little bit inconvenient, but not a huge problem.  I need figure out how to bootstrap Stylet?  Does it have to be attached to a xaml window?  Is it possible that I bootstrap it to an invisible window and let it serve as the root window and all the real windows will be managed by it?

**Reply by canton7** [reply]
> What should we do in a class library project then?

A normal class library project would define the bootstrapper, and the exe project instantiates the ApplicationLoader, referencing the bootstrapper.

The Bootstrapper works by hooking into the Application's Startup event and using that to show the root window. I suggest reading the source of ApplicationLoader and BootstrapperBase - they're pretty simple

What does Office give you to work with? Do you get an Application?

**Reply by mwdpb** [reply]
I read the document one more time, it seems if I can translate this xaml piece into code, and put it in the button call back which invokes GUI, I should be all set.

<Application x:Class="Stylet.Samples.Hello.App"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:s="https://github.com/canton7/Stylet"
             xmlns:local="clr-namespace:Stylet.Samples.Hello">
    <Application.Resources>
       <s:ApplicationLoader>
            <s:ApplicationLoader.Bootstrapper>
                <local:Bootstrapper/>
            </s:ApplicationLoader.Bootstrapper>
        </s:ApplicationLoader>
    </Application.Resources>
</Application>

**Reply by canton7** [reply]
If the Application's Startup event has already fired that probably won't work on its own, and you'll need to manually kick the Bootstrapper to tell it to start.

**Comment by mwdpb** [comment]
![image](https://user-images.githubusercontent.com/55845355/106365206-f31fd980-6301-11eb-9f88-f562e894c0a9.png)

This is what Office VSTO project look like.  Basically it defines a class and subscribe to Office startup and shutdown.  The bookstrapping point is at "ThisAddIn_Startup" call back.

**Reply by canton7** [reply]
Presumably you have access to a System.Windows.Application in there?

Digging into the `BootstrapperBase` code, it seems pointless to call [`Setup`](https://github.com/canton7/Stylet/blob/b2db34a8014efee0bd23d03e965fbc901e6038a4/Stylet/BootstrapperBase.cs#L38): that's setting up stuff which happens when the Application starts up and shuts down, which you probably don't want to be doing here? I guess it depends on whether you want to set everything up once, or every time ThisAddIn_Startup is called.

You might want to do something like (completely untested):

```cs
public class MyBootstrapper : Bootstrapper<ShellViewModel>
{
    public void StartAddIn()
    {
        Setup(Application.Current);
        Start();
    }

    public void StopAddIn()
    {
        Dispose();
    }
}
```

and then call `StartAddIn()` from `ThisAddIn_Startup()`, and `StopAddIn()` from `ThisAddIn_Shutdown()`. However that does leak the bootstrapper as it's retained as a subscriber to the Application's Startup and Exit events: I'd happily take a PR to resolve that.

**Comment by mwdpb** [comment]
FWIW, I was able to use Caliburn.Micro in an previous Office VSTO project.  What I did was in that "ThisAddIn_Startup" callback do the following initialization where ```_cmWindowManager``` is a private field that I can reference anytime.

```
Bootstrapper bootstrapper = new Bootstrapper();
_cmWindowManager = new WindowManager();
```

When I click a button on my addin ribbon, it fires a callback and I can launch GUI using the code below:
```
_cmWindowManager.ShowWindow(new MyViewModel());
```

**Reply by canton7** [reply]
Cool, you can do exactly the same in Stylet: that's probably the easiest way to be honest. `Launch` is probably the method to call, but I strongly encourage you to [read the BootstrapperBase code](https://github.com/canton7/Stylet/blob/b2db34a8014efee0bd23d03e965fbc901e6038a4/Stylet/BootstrapperBase.cs) to decide which bits to call when the bootstrapper is instantiated and which bits to call when your adding is called: it really isn't complicated. All I'm doing at the moment is reading that code and repeating it back to you :)

**Reply by canton7** [reply]
... or just launch the root window yourself with the WindowManager if you'd rather.

**Reply by mwdpb** [reply]
Thanks, I’ll look into that and report back, need to step out for something else now.

**Reply by mwdpb** [reply]
![image](https://user-images.githubusercontent.com/55845355/106367920-35521680-6314-11eb-9cd3-a14fb372f3d5.png)

I tried the same code as CM and it is a no go.  It errored out without specifying a viewmodel as unlike CM, ```Bootstrapper``` in Stylet is a generic class.  It is also an abstract class, so I need inherit it somewhere in my project before I can instantiate it.  The question is where should I do the inheritance.

I'm peeking thru the ```BootstrapperBase``` code, should I cherry pick different pieces to make it work for my case?  What pieces are necessary?

**Reply by canton7** [reply]
Well yes, you're expected to subclass Bootstrapper, as that's where stuff like the ioc container setup happens. You of course need to be instantiating your Bootstrapper subclass.

**Comment by mwdpb** [comment, marked as answer]
So I dug into the ```Bootstrapper``` class and looks like if I don't ```Launch```, I should be able to initialize Stylet without showing any Window, right?  

Here is another question: does rootview bear any importance thru out the program life?  Can I not assign it at all, I mean, get rid of ```_rootViewModel```, ```RootViewModel``` all together?

**Reply by mwdpb** [reply]
I think I'm getting close.  I put the following 3 lines in the VSTO startup event callback:

![image](https://user-images.githubusercontent.com/55845355/106373791-2a66a880-634b-11eb-8541-36808a15afbd.png)

I was able to run it without the third line: I can see message box "Addin started!" which proves the callback did get fired.  The ```Bootstrapper bootstrapper = new BootStrapper();``` gets thru without error.  Now I need figure out how to hook ```bootstrapper``` to ```ApplicationLoader``` in code (the third line is just a mock up, it won't run).  You did that by instantiate them in the ```App.xaml``` as below:

```
<Application.Resources>        
            <s:ApplicationLoader>
                <s:ApplicationLoader.Bootstrapper>
                    <local:Bootstrapper/>
                </s:ApplicationLoader.Bootstrapper>
            </s:ApplicationLoader>        
</Application.Resources>
```

**Reply by mwdpb** [reply]
Here is the ```BootStrapper``` class that's inherited from `Bootstrapper```.  I disabled the ```Launch``` override because I want to see the root view gets displayed first.  Once it is displayed, I'll then uncomment ```Launch``` override and everything should run smoothly.

![image](https://user-images.githubusercontent.com/55845355/106374005-0dcb7000-634d-11eb-9bf6-d2ce0c8850c1.png)

**Reply by canton7** [reply]
I was imaginging you'd do something like this:

```cs
public class AddInBootstrapper : Bootstrapper<ShellViewModel>
{
    protected override void Launch()
    {
        // Don't show the root ViewModel here -- wait for AddInStartup
    }

    public void AddInStartup()
    {
        DisplayRootView(RootViewModel);
    }

    public void AddInShutdown()
    {
        RootViewModel.RequestClose();
    }
}

public class AddIn
{
    private readonly Bootstrapper bootstrapper;

    public AddIn()
    {
        bootstrapper = new Bootstrapper();
        bootstrapper.Setup(Application.Current);
    }

    public void ThisAddIn_Startup(object sender, EventArgs e) => bootstrapper.AddInStartup();
    public void ThisAddIn_Shutdown(object sender, EventArgs e) => bootstrapper.AddInShutdown();
}
```

**Reply by mwdpb** [reply]
Make sense and just tried your snippet.  The problem is at ```Setup(Application.Current)```.  It looks VSTO addin project doesn't support ```Application.Current```.  Here is the error message.

![image](https://user-images.githubusercontent.com/55845355/106389614-2de14a80-63b2-11eb-9113-fef7c5777d18.png)

**Reply by canton7** [reply]
Stylet is going to need an Application instance from somewhere I think (unless you strip out the bits that assume one is present). Is there any other way to get your hands on one?

**Reply by mwdpb** [reply]
VSTO project does provide you a class field called "Application".  You can use it right the way, for example, the following code runs without error:

![image](https://user-images.githubusercontent.com/55845355/106390543-dd202080-63b6-11eb-9152-8047a05d573e.png)

I went to my old VSTO project with CM, its ```Bootstrapper``` looks like this.  I then went to CM ```BootstrapperBase``` and found the ```false``` indicates "do not use Application".  See its source code here: [https://github.com/Caliburn-Micro/Caliburn.Micro/blob/master/src/Caliburn.Micro.Platform/Platforms/net46/Bootstrapper.cs](url)

![image](https://user-images.githubusercontent.com/55845355/106390593-12c50980-63b7-11eb-8bec-abd71ef85630.png)

**Reply by canton7** [reply]
I'd take a PR which removes the dependency on Application

**Reply by mwdpb** [reply]
Thanks, I will.  BTW, do you have any design document I can reference to other than the Wiki page?  I felt I need understand quite a bit of the code before making meaningful changes.

**Reply by canton7** [reply]
There aren't any design docs I'm afraid -- it's only a one-man show!

I was thinking something like 66d68aed3efabf16d2680f5ca02daeba055fa8b1 -- there you can simply not call `Setup` and instead call `Start`

**Reply by mwdpb** [reply]
I submitted a pull request.  What I did is added an abstract class called ```BootstrapperNoRootView``` and have the original ```Bootstrapper``` class inherit from it.  Please take a look to see if that'll break anything.

Back to the original question, I was able to get it running by the following code snippet:

![image](https://user-images.githubusercontent.com/55845355/106829265-cfd69080-6659-11eb-9d51-a4b7f3a8904b.png)

![image](https://user-images.githubusercontent.com/55845355/106829286-d7963500-6659-11eb-9103-5c2f0fc54051.png)

Later when you do need invoke a GUI, use this:

![image](https://user-images.githubusercontent.com/55845355/106829590-715de200-665a-11eb-8a17-6ed3171077cf.png)

Hopefully, this will help others who have similar needs.  :)
